libs/minecraft/CommandUtils.arucas

```diff
@@ -0,0 +1,420 @@
+
+class FILOStack {
+    var asList;
+    FILOStack(fromList){
+        this.asList = [];
+        this.asList.addAll(fromList.copy());
+    }
+    fun popLast(){
+        pointer = len(this.asList);
+        if(pointer == 0){
+            return null;
+        }
+        val = this.asList.get(pointer - 1);
+        this.asList.remove(pointer - 1);
+        return val;
+    }
+    fun pop(){
+        pointer = 0;
+        if(pointer == len(this.asList)){
+            return null;
+        }
+        val = this.asList.get(pointer);
+        this.asList.remove(pointer);
+        return val;
+    }
+    fun hasItems(){
+        return !this.asList.isEmpty();
+    }
+}
+fun tryPredicateAndMerge(iterable, predicate, modifyFunction){
+    retVal = [];
+    foreach ( i : iterable){
+        defaultVal = false;
+        try {
+            defaultVal = predicate(i);
+        }
+        catch (ignored) {
+            continue;
+        }
+        retVal.append(modifyFunction(i));
+    }
+    return retVal;
+}
+
+lambdaItems = fun (obj){
+    return obj.asItemStack() != null;
+};
+lambdaItemGetId = fun (obj){
+    return obj.asItemStack().getId();
+};
+
+lambdaBlock = fun (obj){
+    return obj.asBlock() != null;
+};
+lambdaBlockGetId = fun (obj){
+    return obj.asBlock().getId();
+};
+
+class DefaultSuggester {
+    static var ITEMS = tryPredicateAndMerge(Material.ALL, lambdaItems, lambdaItemGetId);
+    static var BLOCKS= tryPredicateAndMerge(Material.ALL, lambdaBlock, lambdaBlockGetId);
+    static var DIRECTIONS = ['up', 'down', 'north', 'east', 'south', 'west'];
+    static var HORIZONTAL = ['north', 'east', 'south', 'west'];
+    static var BIOMES = ['THE_VOID', 'PLAINS', 'SUNFLOWER_PLAINS', 'SNOWY_PLAINS', 'ICE_SPIKES', 'DESERT', 'SWAMP', 'MANGROVE_SWAMP', 'FOREST', 'FLOWER_FOREST', 'BIRCH_FOREST', 'DARK_FOREST', 'OLD_GROWTH_BIRCH_FOREST', 'OLD_GROWTH_PINE_TAIGA', 'OLD_GROWTH_SPRUCE_TAIGA', 'TAIGA', 'SNOWY_TAIGA', 'SAVANNA', 'SAVANNA_PLATEAU', 'WINDSWEPT_HILLS', 'WINDSWEPT_GRAVELLY_HILLS', 'WINDSWEPT_FOREST', 'WINDSWEPT_SAVANNA', 'JUNGLE', 'SPARSE_JUNGLE', 'BAMBOO_JUNGLE', 'BADLANDS', 'ERODED_BADLANDS', 'WOODED_BADLANDS', 'MEADOW', 'GROVE', 'SNOWY_SLOPES', 'FROZEN_PEAKS', 'JAGGED_PEAKS', 'STONY_PEAKS', 'RIVER', 'FROZEN_RIVER', 'BEACH', 'SNOWY_BEACH', 'STONY_SHORE', 'WARM_OCEAN', 'LUKEWARM_OCEAN', 'DEEP_LUKEWARM_OCEAN', 'OCEAN', 'DEEP_OCEAN', 'COLD_OCEAN', 'DEEP_COLD_OCEAN', 'FROZEN_OCEAN', 'DEEP_FROZEN_OCEAN', 'MUSHROOM_FIELDS', 'DRIPSTONE_CAVES', 'LUSH_CAVES', 'DEEP_DARK', 'NETHER_WASTES', 'WARPED_FOREST', 'CRIMSON_FOREST', 'SOUL_SAND_VALLEY', 'BASALT_DELTAS', 'THE_END', 'END_HIGHLANDS', 'END_MIDLANDS', 'SMALL_END_ISLANDS', 'END_BARRENS'];
+    static fun getAvailableProperties(propName) {
+        immutableMap = {
+            'lit' : ['true', 'false'],
+            'facing' : ['up', 'down', 'north', 'east', 'south', 'west'],
+            'face' : ['wall', 'ceiling', 'floor'],
+            'powered' : ['true', 'false'],
```

Comment:
This map is created every time this function is called. It should be stored as a static variable.